parlai/mturk/core/mturk_manager.py

```diff
@@ -1859,3 +1862,186 @@ def email_worker(self, worker_id, subject, message_text):
+            allowed_convs = self.opt['allowed_conversations']
+            if allowed_convs > 0 and convs >= allowed_convs:
+                text = ('You can participate in only {} of these HITs at '
+                        'once. Please return this HIT and finish your '
+                        'existing HITs before accepting more.'.format(
+                            allowed_convs
+                        ))
+                self.force_expire_hit(worker_id, assign_id, text)
+                return
+
+            # Initialize a new agent for this worker
+            self.worker_manager.assign_task_to_worker(
+                hit_id, assign_id, worker_id
+            )
+            if self.db_logger is not None:
+                self.db_logger.log_worker_accept_assignment(
+                    worker_id, assign_id, hit_id)
+            agent = self.worker_manager._get_agent(worker_id, assign_id)
+            self._add_agent_to_pool(agent)
+            # TODO agents are added directly to the pool, but their liveliness
+            # isn't monitored. Start spawining something to kill a world
+            # thread by marking an agent as disconnected if they've been
+            # on for longer than the task duration.
+        else:
+            # Reconnecting worker
+            agent = self.worker_manager._get_agent(worker_id, assign_id)
+            agent.log_reconnect()
+            agent.alived = True
+            if agent.conversation_id is not None and \
+                    conversation_id is not None:
+                # agent.conversation_id is None is used in testing.
+                # conversation_id is None on a fresh reconnect event, where
+                # we need to restore state somehow and shouldn't just inherit
+                conversation_id = agent.conversation_id
+            if agent.get_status() == AssignState.STATUS_NONE:
+                agent.set_status(AssignState.STATUS_IN_TASK)
+                return
+            elif agent.get_status() == AssignState.STATUS_WAITING:
+                if self.is_task_world(conversation_id):
+                    agent.set_status(AssignState.STATUS_IN_TASK)
+                    return
+                # Reconnecting in waiting is either the first reconnect after
+                # being told to wait or a waiting reconnect. Restore state if
+                # no information is held, and add to the pool if not already in
+                # the pool
+                if not conversation_id:
+                    self._restore_agent_state(worker_id, assign_id)
+            elif agent.get_status() == AssignState.STATUS_IN_TASK:
+                # Reconnecting to the onboarding world or to a task world
+                # should resend the messages already in the conversation
+                if not conversation_id:
+                    self._restore_agent_state(worker_id, assign_id)
+            elif (agent.get_status() == AssignState.STATUS_DISCONNECT or
```

Comment:
this or statement might be good as a helper

Reply:
I actually already have this as a helper (`agent.is_final()`), but I didn't want to obfuscate the switch-case like flow that this if block is demonstrating. This logic should be reduced in the v1 implementation down the road anyways though.